File: DebrisFromExercises/ch11json/JsonExample.py
import sys
import json

# Name of the Members
# Everything:
Token = 'Token'  # the token type

# Terminal:
Value = 'Value'
# Identifier:
Type = 'Type'
Kind = 'Kind'
Index = 'Index'
# SubroutineDec
VarCount = 'VarCount'
# SubroutineDec - constructor
FieldCount = 'FieldCount'

# NonTerminal
Children = 'Children'

def writeLine(line):
    print line
    
def readFileText():
    jsonFile = sys.argv[1]
    writeLine("Attempting to open file: " + jsonFile)
    fileText = open(jsonFile).read()
    writeLine("Read " + str(len(fileText)) + " characters.")
    return fileText


def codeSubroutineDec(subroutineDec):
    children = subroutineDec[Children]
    subroutineType = children[0][Value]
    type = children[1][Value]
    name = children[2][Value]
    localCount = subroutineDec[VarCount]

    if subroutineType == 'constructor':
        fieldCount = subroutineDec[FieldCount]
        suffix = ', oh and the class has ' + fieldCount + ' fields.'
    else:
        suffix = '.'
    line = '    ' + subroutineType.upper() + ' "' + name + '" returns a ' + type + \
        ', it has ' + localCount + ' local variables' + suffix
    writeLine(line)
    #  we don't need to do anything with the parameter list (used by symbol table)


def codeClass(_class):
    children = _class[Children]
    className = children[1][Value]
    writeLine('')
    writeLine("I'm in CLASS " + className)
    for child in children:
        if child[Token] == 'subroutineDec':
            codeSubroutineDec(child)
    #  N.B.
    #  We don't have to do anything with the classVarDecs.
    #  that information was used by the symbol table.
    writeLine("Ding, ding, end of CLASS " + className)
    writeLine('')


def main():
    fileText = readFileText()
    tree = json.loads(fileText)
    codeClass(tree)
  
main()

File: DebrisFromExercises/ch11json/JsonExample.cs
using Newtonsoft.Json;
using System;
using System.IO;
using System.Text;

namespace JackCompiler
{
    // Just to demonstrate how to consume the JSON tree in C#
    class JsonExample
    {
        //Name of the Members
        //  Everything:
        //      Token   # the token type
        //
        //  Terminal:
        //      Value
        //  Identifier:
        //      Type
        //      Kind
        //      Index
        //  SubroutineDec
        //      VarCount
        //  SubroutineDec - constructor
        //      FieldCount
        //
        //  NonTerminal
        //      Children

        public void CodeFile(string filePath)
        {
            string fileText;
            using (var reader = File.OpenText(filePath))
            {
                fileText = reader.ReadToEnd();
            }
            WriteLine();
            CodeFileText(fileText);
        }

        public void CodeFileText(string fileText)
        {
            dynamic tree = JsonConvert.DeserializeObject(fileText);
            var children = tree.Children;
            if (tree.Token != "class")
                throw new Exception("Expected a class");
            CodeClass(tree);
        }

        void CodeClass(dynamic tree)
        {
            var children = tree.Children;
            var className = children[1].Value;
            WriteLine("I'm in CLASS " + className);
            
            foreach (var child in children)
            {
                if (child.Token == "subroutineDec")
                    CodeSubroutineDec(child);
            }
            /*  N.B.
             *  We don't have to do anything with the classVarDecs.
             *  that information was used by the symbol table.
             */
            WriteLine("Ding, ding, end of CLASS ", className);
        }

        void CodeSubroutineDec(dynamic subroutineDec)
        {
            var children = subroutineDec.Children;
            var subroutineType = (string)children[0].Value;
            var type = children[1].Value;
            var name = children[2].Value;
            var localCount = subroutineDec.VarCount;

            string suffix = ".";
            if (subroutineType == "constructor")
            {
                var fieldCount = subroutineDec.FieldCount;
                suffix = ", oh and the class has " + fieldCount + " fields.";
            }
            WriteLine("    ", subroutineType.ToUpper(), " '", name, "' returns a ", type, ", it has ",
                localCount, " local variables", suffix);
            /* N.B.  we don't need to do anything with the parameter list */
        }

        void WriteLine(string line)
        {
            Console.WriteLine(line);
        }

        void WriteLine(params object[] parts)
        {
            WriteLine(string.Concat(parts));
        }

    }
}
